### PR TITLE
Change fetchMetadata return type for interfaces and objects to be  type definition

### PR DESCRIPTION
**Pull request by nihalbhatnagar**
This PR accomplishes 3 things

- Implements MinDef into interfaces and moves the definition of generated interface types to the root
- Creates a new abstraction for Object and Interface types that allows for the separation of compile time metadata and metadata that we should fetch
- Implements the new types into fetch metadata

### CODE REVIEW

File: packages/generator/src/v2.0/UNSTABLE_wireInterfaceTypeV2ToSdkObjectConst.ts [hunk 1]
```diff
@@ -175,56 +175,25 @@ ${
...
   const imports: string[] = [];
-
+  definition;
   return `${imports.join("\n")}
     ${v2 ? getV2Types() : ""}
 
-    export const ${interfaceDef.shortApiName}: ${interfaceDef.shortApiName}.Definition = {
+    export const ${interfaceDef.shortApiName}: ${interfaceDef.shortApiName} = {
+      type: "interface",
+      apiName: "${interfaceDef.fullApiName}",
```
Comment (nihalbhatnagar) [inline packages/generator/src/v2.0/UNSTABLE_wireInterfaceTypeV2ToSdkObjectConst.ts:196]: We do type: "${object instanceof EnhancedObjectType ? "object" : "interface"}" in object. Is this needed here if no errors?

Comment (ericanderson) [reply packages/generator/src/v2.0/UNSTABLE_wireInterfaceTypeV2ToSdkObjectConst.ts:196]: If we do it there and its just for objects then its probably a mistake where I thought those lines were being reused

**Review by ericanderson — CHANGES_REQUESTED** [review]

File: packages/api/src/ontology/ObjectTypeDefinition.ts [hunk 2]
```diff
@@ -33,19 +33,14 @@ export type ObjectTypePropertyDefinitionFrom2<
 export interface ObjectInterfaceBaseDefinition<K extends string, N = unknown> {
   type: "object" | "interface";
   apiName: K;
-  displayName?: string;
+  displayName: string;
```
[the remark this answers is not in the document]
Comment (nihalbhatnagar) [reply packages/api/src/ontology/ObjectTypeDefinition.ts:36]: We always get a value from the backend, unless you're implying that they do this replacement for us
`export interface ObjectTypeV2 {
  apiName: ObjectTypeApiName;
  displayName: DisplayName; `

File: etc/client.report.api.md [hunk 3]
```diff
@@ -78,7 +76,7 @@ export interface Client extends SharedClient<MinimalClient> {
     // (undocumented)
     <Q extends MinQueryDef<any, any, any>>(o: Q): QuerySignatureFromDef<Q>;
     // (undocumented)
-    fetchMetadata<Q extends (MinObjectDef<any, any> | MinInterfaceDef<any, any> | MinActionDef<any, any, any> | MinQueryDef<any, any, any>)>(o: Q): Promise<Q extends MinObjectDef<any, any> ? ObjectMetadata : Q extends MinInterfaceDef<any, any> ? InterfaceMetadata : Q extends MinActionDef<any, any, any> ? ActionMetadata : Q extends MinQueryDef<any, any, any> ? QueryMetadata : never>;
+    fetchMetadata<Q extends (MinObjectDef<any, any> | MinInterfaceDef<any, any> | MinActionDef<any, any, any> | MinQueryDef<any, any, any>)>(o: Q): Promise<Q extends MinObjectDef<any, any> ? ObjectTypeDefinition<any, any> : Q extends MinInterfaceDef<any, any> ? InterfaceDefinition<any, any> : Q extends MinActionDef<any, any, any> ? ActionMetadata : Q extends MinQueryDef<any, any, any> ? QueryMetadata : never>;
```
[the remark this answers is not in the document]
Comment (nihalbhatnagar) [reply etc/client.report.api.md:79]: Will do in a FLUP, change action and queries and delete our metadata types. Will also rename definition

File: packages/api/src/ontology/ObjectTypeDefinition.ts [hunk 4]
```diff
@@ -64,42 +68,35 @@ export interface ObjectTypeDefinition<
...
   primaryKeyApiName: keyof this["properties"];
+  titleProperty: keyof this["properties"];
   primaryKeyType: PrimaryKeyTypes;
-
-  /**
-   * Optional because they may not exist on legacy.
-   */
-  interfaceMap?: Record<
+  icon?: Icon;
+  visibility?: ObjectTypeVisibility;
```
[the remark this answers is not in the document]
Comment (nihalbhatnagar) [reply packages/api/src/ontology/ObjectTypeDefinition.ts:74]: I can double check with a backend team, but this doesn't necessarily come back on our type: `
export interface ObjectTypeV2 {
  apiName: ObjectTypeApiName;
  displayName: DisplayName;
  status: ReleaseStatus;
  description?: string;
  pluralDisplayName: string;
  icon: Icon;
  primaryKey: PropertyApiName;
  properties: Record<PropertyApiName, PropertyV2>;
  rid: ObjectTypeRid;
  titleProperty: PropertyApiName;
  visibility?: ObjectTypeVisibility; `. Do you know who I could talk to?

File: packages/api/src/ontology/ObjectTypeDefinition.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ericanderson) [reply packages/api/src/ontology/ObjectTypeDefinition.ts:36]: in the earlier versions it could be undefined thats why im asking. didnt know they changed it

File: packages/api/src/ontology/ObjectTypeDefinition.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ericanderson) [reply packages/api/src/ontology/ObjectTypeDefinition.ts:74]: Cool, just checking

File: etc/client.report.api.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (nihalbhatnagar) [reply etc/client.report.api.md:79]: So currently we have 
```export interface ActionDefinition<
  A extends string,
  K extends string,
> {
  type: "action";
  apiName: A;
  description?: string;
  displayName?: string;
  parameters: Record<any, ActionParameterDefinition<K, any>>;
  modifiedEntities?: Partial<Record<K, ActionModifiedEntity>>;
}

export interface ActionDefMetadata<A extends string, K extends string, T>
  extends ActionDefinition<A, K>
{
  signatures: T;
}

export interface MinActionDef<``` in actions. The "metadata" for runtime should be the ActionDefinition, right? So we need to rename this and I think we should follow the []CompileDefinition

File: packages/generator/src/v2.0/UNSTABLE_wireInterfaceTypeV2ToSdkObjectConst.ts (hunk 1 above)
Comment (nihalbhatnagar) [reply packages/generator/src/v2.0/UNSTABLE_wireInterfaceTypeV2ToSdkObjectConst.ts:196]: Cool will fix in flup

File: packages/client/src/ontology/OntologyProvider.ts [hunk 5]
```diff
@@ -28,8 +28,6 @@ export const InterfaceDefinitions = Symbol(
 export interface FetchedObjectTypeDefinition<K extends string, N = unknown>
   extends ObjectTypeDefinition<K, N>
 {
-  rid: string;
```
Comment (ericanderson) [inline packages/client/src/ontology/OntologyProvider.ts:31]: Why this change

**Review by ericanderson — APPROVED** [review]
Approved, looking forward to the followups
